content/en/blog/_posts/2024-04-15-devsecops-in-action-k8s-secrets.md

```diff
@@ -0,0 +1,535 @@
+---
+layout: blog
+title: "DevSecOps in Action: Kubernetes Secrets"
+slug: devsecops-in-action-k8s-secrets
+date: 2024-04-15
+---
+
+**Author**: Evgenii Frikin
+
+The Secret is one of the basic resources that exists in the Kubernetes API. This
+kind of resource is used for different purposes and despite the name Secret,
+sensitive data is insecure to store in it. Recently considerable attention has
+been paid to
+[security improvement](/docs/tasks/administer-cluster/encrypt-data/)
+for kind of Secret resource, but Kubernetes still doesn't include appropriate
+out-of-the-box because Kubernetes philosophy carries simplicity in order not to
+compromise it engineers have to make sure Kubernetes Secret are saved
+
+According to
+[Good practices for Kubernetes Secrets](/docs/concepts/security/secrets-good-practices/)
+there are several options on how to improve security of this kind of Secret:
+encrypt secrets in the etcd, restrict the access to them via RBAC, configure
+access to external Secret, etc. However this partially solves the security issue
+because they are cluster side solutions and they help to improve security of the
+cluster overall.
+
+In the modern world unilateral solutions aren't effective enough and can hide
+security issues known as
+[Security through obscurity](https://en.m.wikipedia.org/wiki/Security_through_obscurity).
+In order to avoid it it's necessary to integrate
+[DevSecOps methodology](https://www.redhat.com/en/topics/devops/what-is-devsecops)
+which implies integrated security as a shared responsibility throughout the
+entire IT lifecycle. In other words, security solutions must start with the
+development process and apply to all other levels correspondingly.
+
+According to
+[OWASP Secrets Management Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Secrets_Management_Cheat_Sheet.html#5-containers-orchestrators)
+it's recommended to use Injection of Secret via file and in-memory. Provision of
+sensitive data via environment variables isn't recommended, but at the same time
+it's recommended in The
+[Twelve Factors](https://12factor.net/config).
+These are general recommendations which don't impact security issues that's why
+their application should make sense,but passing config via environment variables
+with sensitive data must be kept to a minimum.
+
+Injection of Secret via file and in-memory has pros and cons and I would like to
+show which security issues exist when using Injection of Secret via file and
+suggest solutions. In fact, there will be a general solution and specific
+implementation will be different in terms of specific management and delivery
+secrets tools. To keep the examples simple I will use `spec.initContainers` to
+simulate an injection agent for delivering secrets to Pod.
+
+## Baseline
+
+Imagine there is a simple application which processes http requests. On request
```

Comment:
```suggestion
Imagine there is an application which processes http requests. On request
```
Avoided a word that [assumes](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-words-that-assume-a-specific-level-of-understanding) a specific level of understanding.

Reply:
The original text is actually find @efrikin - you're describing an application that is not complex.

Reply:
I appreciate your feedbacks, @sftim.